### PR TITLE
PyTorch RCFR Implementation

### PR DESCRIPTION
**Pull request by Asugawara**
Adding RCFR algorithms replacing TensorFlow to PyTorch.(Almost same interface.)
It will be helpful for the community. And this is mentioned in the [documentation.](https://openspiel.readthedocs.io/en/latest/contributing.html#roadmap-and-call-for-contributions)

This PR have some experiment in [notebook](https://github.com/deepmind/open_spiel/blob/ca4c3f72a267c4d44208bc56b0761f73076b70ef/open_spiel/colabs/rcfr_pytorch.ipynb). Please check it.
Compare exploitability with TensorFlow and PyTorch.
 
Thanks for your development. Open_spiel is very cool oss!!

### CODE REVIEW

**Comment by Asugawara** [comment]
@googlebot I fixed it.

**Comment by Asugawara** [comment]
@googlebot I fixed it.

**Comment by lanctot** [comment]
Thanks @Asugawara!

That is quite a lot of new dependencies. Can you explain why you need this many for an RCFR implementation? We generally prefer to keep OpenSpiel as light as possible, and dependencies add long-term maintenance burdens.

The nox tests on Travis-CI are breaking as a result, any idea what is causing that?

**Comment by Asugawara** [comment]
Thanks for response @lanctot !

This pullrequest added only torch and hypothesis.
torch is PyTorch.
PyTorch’s TestCase class imports hypothesis which uses in rcfr_pytorch_test.py replacing tf.test.TestCase.
(ref:https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L678)

typing_extensions、future,  numpy,  dataclasses are PyTorch's requirements.
(ref:https://github.com/pytorch/pytorch/blob/master/requirements.txt)

attrs, sortedcontainers are hypothesis’s requirements.
(ref:https://github.com/HypothesisWorks/hypothesis/blob/e2db37d0d8c0b2981f87670f92e13cafaa58ba78/hypothesis-python/setup.py#L96)

So, I fix requirements.txt. and the nox tests on Travis-CI will go smoothly.

(Breaking the nox tests on Travis-CI caused by importing dataclass lib)

**Comment by lanctot** [comment]
> This pullrequest added only torch and hypothesis.

Huh, for some reason I saw many others added to requirements.txt the last time I looked. Maybe it's because I was looking at an individual commit, or maybe you changed it since. Adding these two is ok :)

**Comment by lanctot** [comment]
> So, I fix requirements.txt. and the nox tests on Travis-CI will go smoothly.
> 
> (Breaking the nox tests on Travis-CI caused by importing dataclass lib)

Ahh, ok, there you go! Perfect.

Ok, we will take a look... Thanks again for the contribution!

(Tagging @dmorrill10 and @michalsustr who might be interested.)

**Comment by lanctot** [comment]
Let me throw @ssokota into the tag mix as he has more PyTorch experience than anyone I know who is familiar with CFR-style algorithms.

File: open_spiel/python/pytorch/rcfr.py [hunk 1]
```diff
@@ -0,0 +1,835 @@
+from __future__ import absolute_import
```
Comment (lanctot) [inline open_spiel/python/pytorch/rcfr.py:1]: At the very top of the file, san you add some text here describing RCFR? Does not have to be long (one or two sentences), but I would also like you to link to the original paper and Dustin's thesis. 

(You can just grab most of the text comment in the TF rcfr.py)

**Review by lanctot — CHANGES_REQUESTED** [review]

**Review by lanctot — APPROVED** [review]

**Review by lanctot — CHANGES_REQUESTED** [review]

File: open_spiel/colabs/rcfr_pytorch.ipynb [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
...tqFewn09zhdDXw+Rvm4ifb3Yvh6ioR5coV4C4V4ucKdNRFWnUVCLyNTl9joX/+jtCgN4Y5d4GN+21SIict4pqEVEEpyCWkQkwSmoRUQSnIJaRCTBKahFRBKcglpEJMEpqEVEEtyk7JlojGkCDpzjy6cBJyawnGSmz2I4fR7D6fMYlAqfRZm1tnCkByYlqN8NY8yWM23wONXosxhOn8dw+jwGpfpnodaHiEiCU1CLiCS4RAzqn8S7gASiz2I4fR7D6fMYlNKfRcL1qEVEZLhEPKIWEZEhFNQiIgkuYYLaGPNBY8weY8w7xph74l1PPBljSowxzxtj3jLG7DTGfDHeNcWbMcZtjHnNGPPHeNcSb8aYXGPM48aY3caYXcaYy+NdUzwZY74c++9khzHmV8aYQLxrmmgJEdTGGDfwI+AaYD7wMWPM/PhWFVdh4KvW2vnACuCvp/jnAfBFYFe8i0gQ/wz8l7V2LrCIKfy5GGNmAV8AllprKwE38NH4VjXxEiKogWXAO9ba/dbaPuA/gP8R55rixlp71Fq7LfZzB85/iLPiW1X8GGOKgb8CfhbvWuLNGJMDrAYeBrDW9llrW+NbVdx5gDRjjAdIB47EuZ4JlyhBPQs4NOR2A1M4mIYyxpQDS4BN8a0krh4E/gbexZbRqaMCaALWxVpBPzPGjG932BRirT0MfA84CBwF2qy1z8S3qomXKEEtIzDGZAK/Bb5krW2Pdz3xYIy5Fmi01m6Ndy0JwgNUAQ9Za5cAXcCUPadjjMnD+eu7ApgJZBhj1sa3qomXKEF9GCgZcrs4dt+UZYzx4oT0Y9ba38W7njhaCVxvjKnHaYmtMcY8Gt+S4qoBaLDW9v+F9ThOcE9V7wPqrLVN1toQ8DvgPXGuacIlSlBvBi42xlQYY3w4JwP+M841xY0xxuD0IHdZax+Idz3xZK39urW22FpbjvP/i+estSl3xDRW1tpjwCFjzJzYXVcBb8WxpHg7CKwwxqTH/ru5ihQ8ueqJdwEA1tqwMeZu4L9xztr+3Fq7M85lxdNK4FbgTWPM67H7/tZa+1Qca5LE8XngsdhBzX7gjjjXEzfW2k3GmMeBbTijpV4jBaeTawq5iEiCS5TWh4iInIGCWkQkwSmoRUQSnIJaRCTBKahFRBKcglpEJMEpqEVEEtz/B9BFl9Iw4HeMAAAAAElFTkSuQmCC\n",
```
[the remark this answers is not in the document]
Comment (Asugawara) [reply open_spiel/colabs/rcfr_pytorch.ipynb:204]: Hi! @lanctot !  Yes I could, those images are not important.
 I remove it.
Thank you for your pointing out.

[the remark this answers is not in the document]
Comment (lanctot) [reply open_spiel/colabs/rcfr_pytorch.ipynb]: Great, thanks!

**Comment by lanctot** [comment]
Hi @Asugawara , from Travis-CI I noticed there's a dependency problem.. seems like one of PyTorch 1.7's dependencies has problems with Python 3.6. 

Can I ask you to try changing `OS_PYTHON_VERSION` from 3.6 to 3.7 in `.travis.yml`? I'd like to do it in this PR to initiate another integration test to see if it passes. 

I noticed Travis also supports Ubuntu 20.04 now, so I should add it as well.

**Review by lanctot — CHANGES_REQUESTED** [review]
The tests are failing. See https://github.com/deepmind/open_spiel/pull/425#issuecomment-726949168 for a suggestion to fix them.

**Comment by Asugawara** [comment]
Thanks. Yes, I will fix it!!!

> seems like one of PyTorch 1.7's dependencies has problems with Python 3.6.
> Can I ask you to try changing OS_PYTHON_VERSION from 3.6 to 3.7

That's what I think, too.

**Comment by lanctot** [comment]
Perfect. Pushing a new commit will trigger another Travis-CI build so we can check if it worked.

**Comment by Asugawara** [comment]
In the above problem,  it was occurred by incompatible between dm-haiku requirements and torch requirements.
```
dataclasses==0.7 ; python_version<'3.7'
https://github.com/deepmind/dm-haiku/blob/master/requirements-test.txt#L7

dataclasses; python_version<“3.7"
https://github.com/pytorch/pytorch/blob/master/requirements.txt#L8

FYI: the latest version is 0.8
```

But, The nox tests on Travis-CI are breaking with the error message saying “An error occurred while generating the build script.”
Do you have any ideas?Travis’s issue?

**Comment by lanctot** [comment]
Yeah that is my guess. I restarted the Travis build, let's see if it was just a fluke.

**Review by lanctot — APPROVED** [review]

**Comment by Asugawara** [comment]
Thank you Mr. @lanctot !

**Comment by lanctot** [comment]
> Thank you Dr. @lanctot !

Very nicely done and well tested, partly benefitting from the TF impl, but still very well adspted.

This is the first officially imported PyTorch PR. A milestone for the project, congrats!

**Pull request merged**